### PR TITLE
Fix cluster desync

### PR DESCRIPTION
**Pull request by Shiranuit**
## What does this PR do ?

~~See #2311 for performances with mutexes~~

Fix https://github.com/kuzzleio/kuzzle/issues/2296

### A bit of background to help understand what does this PR do

Function calls that are awaited causes the event loop to switch to another function and alternate which function is running.
This can cause issues when operations needs to be atomic (done without interruption) like in this case:
**Example of issue using promises where operations needs to be atomic**
```js
let rooms = {}

async function createRoom(room) {
  rooms[room] = {
    subCount: 0
  };
  // After returning from createRoom, the event loop is going to pause the execution of this function
  // and is going to switch to `unusbscribe`.
}

async function subscribe(room) {
  // Call createRoom
  await createRoom(room);
  // `Unsubscribe` is executed
  // After returning from `unsubscribe` the event loop is going to continue executing `subscribe` where it stopped.
  await _subscribe(room); // Call `_subscribe`
  console.log('After Subscribe', rooms[room]);
}

async function _subscribe(room) {
  rooms[room].subCount++;
  // Returning from `_subscribe` the event loop has no other function to execute so it returns back to `subscribe`
}

async function unsubscribe(room) {
  // Once we are here the room has been created but we did not call `_subscribe` yet, so the counter is still 0.
  if (rooms[room]) {
    rooms[room].subCount--;
    if (rooms[room].subCount === 0) {
      delete rooms[room];
    }
  }
  console.log('After Unsubscribe', rooms[room]);
  // After returning from `unsubscribe` the event loop is going to resume the execution of `subscribe`
}

async function main() {
  subscribe(); // Push the subscribe function to the event loop.
  unsubscribe(); // Push the unsubscribe function to the event loop.
  // After returning from main start executing `subscribe`
}

main();
// Printed result:
// After Unsubscribe { subCount: -1 }
// After Subscribe { subCount: 0 }
```

**Example of the event loop not switching to another function until one of the await function of the callstack returns**
```js
async function stepA() {
  console.log('Start stepA');
  // Call `stepB`
  await stepB();
  // Returning from `otherB`
  console.log('End stepA');
  // After returning from `stepA` the event loop is going to resume `otherA`
}

async function stepB() {
  console.log('Start stepB');
  console.log('End stepB');
  // Returning from awaited `stepB`, the event loop is going to switch to `otherA`
}

async function otherA() {
  console.log('Start otherA');
  // Call `otherB`
  await otherB();
  // Returning from `stepA`
  console.log('End otherA');
  // After returning from `otherA` the event loop is going end program since there is no other function to switch to
}

async function otherB() {
  console.log('Start otherB');
  console.log('End otherB');
  // Returning from `otherB` the event loop is going to switch to `stepA` after calling `stepB`
}


async function main() {
  stepA(); // Push stepA to the event loop
  otherA(); // Push otherA to the event loop
  // After returning from main start executing `stepA`
}

main();
// Printed results:
// Start stepA
// Start stepB
// End stepB
// Start otherA
// Start otherB
// End otherB
// End stepA
// End otherA
```

### Changes

To ensure that the subscribe and unsubscribe operations are atomic I needed to reorder and modify some instructions:
- Added `kuzzle.onCall` and `kuzzle.call` to register callback to an event name and calling it synchronously
- Try to move `awaited` function as much as possible after operations that needs to be atomic, i.e: room creation/deletion, subscription to room, unsubscription from room, notification of nodes...
- Replace event that were subscribing / unsubscribing with an async event to a sync callback using `kuzzle.call` instead of `kuzzle.emit`
- Replace some pipe call to `core:realtime:unsubscribe:after` with a sync call using `kuzzle.call` (Internal pipe not documented, should not be breaking change)
- Modify execution order to continue descending the callstack without returning, this way the event loop is not switching back to another function
- Once a node is receiving messages from another node, it now refreshes the heartbeat timeout even though no heartbeat messages has been received, because messages from other nodes should also count has an heartbeat round.

### Benchmark

[See Benchmark](https://github.com/kuzzleio/kuzzle/pull/2310#issuecomment-1094874413)

### How should this be manually tested?

Spam subscription and unsubscriptions on same room

### CODE REVIEW

**Comment by Shiranuit** [comment]
To do: Benchmark changes on cluster subscriptions

**Comment by Shiranuit** [comment]
# Benchmark

## Before

```shell
===============================================================================
=== Real-time control step, measuring raw Kuzzle response time
100% | Requests completed: 10000 | Elapsed: 1s | 8244 requests/s | 0 notifications received
===============================================================================
=== Geopolygon subscriptions listening to geopoints in France
Starting subscriptions...
100 subscriptions done (took: 70ms)
100% | Requests completed: 10000 | Elapsed: 1s | 12300 requests/s | 0 notifications received
Unsubscribing...
Unsubscriptions done (took: 291ms)
===============================================================================
=== Geopolygon subscriptions listening to geopoints in the US
Starting subscriptions...
100 subscriptions done (took: 63ms)
100% | Requests completed: 10000 | Elapsed: 1s | 8097 requests/s | 20000 notifications received
Unsubscribing...
Unsubscriptions done (took: 248ms)
===============================================================================
=== Writing documents to Elasticsearch 1 by 1 (no listeners)
100% | Requests completed: 10000 | Elapsed: 6s | 1720 requests/s | 19600 notifications received
===============================================================================
=== Writing documents to Elasticsearch 1 by 1 (with listeners)
Starting subscriptions...
100 subscriptions done (took: 55ms)
100% | Requests completed: 10000 | Elapsed: 7s | 1424 requests/s | 379962 notifications received
Unsubscribing...
Unsubscriptions done (took: 658ms)
Disconnecting clients: 1000 / 1000
```

## After

```shell
===============================================================================
=== Real-time control step, measuring raw Kuzzle response time
100% | Requests completed: 10000 | Elapsed: 1s | 8326 requests/s | 0 notifications received
===============================================================================
=== Geopolygon subscriptions listening to geopoints in France
Starting subscriptions...
100 subscriptions done (took: 66ms)
100% | Requests completed: 10000 | Elapsed: 1s | 12453 requests/s | 0 notifications received
Unsubscribing...
===============================================================================
=== Geopolygon subscriptions listening to geopoints in the US
Starting subscriptions...
100 subscriptions done (took: 70ms)
100% | Requests completed: 10000 | Elapsed: 1s | 7128 requests/s | 20000 notifications received
Unsubscribing...
Unsubscriptions done (took: 248ms)
===============================================================================
=== Writing documents to Elasticsearch 1 by 1 (no listeners)
100% | Requests completed: 10000 | Elapsed: 6s | 1719 requests/s | 18400 notifications received
===============================================================================
=== Writing documents to Elasticsearch 1 by 1 (with listeners)
Starting subscriptions...
100 subscriptions done (took: 51ms)
100% | Requests completed: 10000 | Elapsed: 9s | 1156 requests/s | 390139 notifications received
Unsubscribing...
```

## Conclusion

Those changes should have no impact, the slight differences between results is not caused by the changes, they are just within the error margin.

File: package.json [hunk 1]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "kuzzle",
   "author": "The Kuzzle Team <support@kuzzle.io>",
-  "version": "2.17.6",
+  "version": "2.17.8",
```
Comment (Aschen) [inline package.json:4]: package-lock.json has not been updated?

Comment (Shiranuit) [reply package.json]: Right I just changed the version for my tests to be sure that I was not using the wrong version of Kuzzle

File: lib/cluster/subscriber.js [hunk 2]
```diff
@@ -216,6 +216,9 @@ class ClusterSubscriber {
     }
 
     try {
+      // If we are receiving messages from a node, 
+      // it means the node is alive so it should counts as an heartbeat
+      this.handleHeartbeat();
```
Comment (Aschen) [inline lib/cluster/subscriber.js:221]: Missing unit test

**Review by Aschen — APPROVED** [review]

**Pull request merged**